File: luaServices/cpu.lua
types = {"collector", "enhancer"}

cfg = {}
cfg["name"] = "cpu"
cfg["poolfreq"] = 10
cfg["history"]  = 150


function declare ()
	this.addRessource("shed", "Service scheduler information", "service_cpu_shed")
	this.addProperty( "shed", "exec_runtime",  "exec_runtime", "number")
	this.addProperty( "shed", "process",  "Number of process", "number")
	this.addProperty( "shed", "thread",  "total number of threads", "number")
	this.addRessource("stat", "Service CPU usage information", "service_cpu_stat")
	this.addProperty( "stat", "user",    "user %",   "number")
	this.addProperty( "stat", "system",  "system %", "number")
	this.addProperty( "stat", "pct",  "total % used", "number")
	this.addProperty( "stat", "cuser",   "user % cumulated",   "number")
	this.addProperty( "stat", "csystem", "system % cumulated", "number")
end

values = {}
users  = {}
sys    = {}
cusers = {}
csys   = {}
prev_total_time = 0

function collect ()
	local f
	local t
	local i          = 1
	local sum        = 0.0
	local process    = 0
	local thread     = 0
	local total_time = 0
	local cpu_count  = 0
	local res_users  = 0
	local res_sys    = 0
	local res_cusers = 0
	local res_csys   = 0
	-- get total_time 1st
	f = assert(io.open("/proc/stat", "r"))
	for line in f:lines() do
		if (line:match("(%a+%d*)") == "cpu") then
			for part in string.gmatch(line, "%d+") do
				total_time  = total_time + tonumber(part)
			end
		elseif (line:match("(%a+)") == "cpu") then
			cpu_count=cpu_count+1
		end
	end
	f:close()
	if (cpu_count==0) then
		cpu_count=1
	end
	this.getPIDList()
	while (type(pids[i]) == "number")
	do
		f = io.open("/proc/".. pids[i] .."/sched", "r")
		if f ~= nill then
			for line in f:lines() do
				if (line:match("(%a+.%a+_%a+_%a+)") == "se.sum_exec_runtime") then
					val = line:match("(%d+.%d+)")
					if(type(values[pids[i] .. ""]) == "number") then
						sum = sum + tonumber(val) - values[pids[i] .. ""]
					end
					values[pids[i] .. ""] = tonumber(val)
				elseif (line:match("threads") == "threads") then
					thread = thread + tonumber(line:match(".*: (%d+)"))
				end
			end
			f:close()
		end
		f = io.open("/proc/".. pids[i] .."/stat", "r")
		if f~=nil then
			t = f:read("*all")
			f:close()
		else
			t = ""
		end
		local n = 3
		for part in string.gmatch(t, "%d+") do
			if n == 14 then
				if(type(users[pids[i] .. ""]) == "number") then
					res_users = res_users+tonumber(part)-users[pids[i] .. ""]
				end
				users[pids[i] .. ""] = tonumber(part)
			elseif n == 15 then
				if(type(sys[pids[i] .. ""]) == "number") then
					res_sys = res_sys+tonumber(part)-sys[pids[i] .. ""]
				end
				sys[pids[i] .. ""] = tonumber(part)
			elseif n == 16 then
				if(type(cusers[pids[i] .. ""]) == "number") then
					res_cusers = res_cusers+tonumber(part)-cusers[pids[i] .. ""]
				end
				cusers[pids[i] .. ""] = tonumber(part)
			elseif n == 17 then
				if(type(csys[pids[i] .. ""]) == "number") then
					res_csys = res_csys+tonumber(part)-csys[pids[i] .. ""]
				end
				csys[pids[i] .. ""] = tonumber(part)
			end
			n=n+1
		end
		i=i+1
		process=process+1
	end
	if prev_total_time ~= 0 then
		this.nextValue("shed")
		this.setDProperty("shed", "exec_runtime", sum)
		this.setDProperty("shed", "process", process)
		this.setDProperty("shed", "thread", thread)
		this.nextValue("stat")
		this.setDProperty("stat", "user",    res_users*100*cpu_count/(total_time-prev_total_time))
		this.setDProperty("stat", "system",  res_sys*100*cpu_count/(total_time-prev_total_time))
		this.setDProperty("stat", "pct",  (res_sys+res_users)*100*cpu_count/(total_time-prev_total_time))
		this.setDProperty("stat", "cuser",   res_cusers*100*cpu_count/(total_time-prev_total_time))
		this.setDProperty("stat", "csystem", res_csys*100*cpu_count/(total_time-prev_total_time))
	end
	prev_total_time = total_time
end

function enhance(serv)
	-- adding this collector to every service
	serv.addCollector("cpu")
end

File: services/cpu.cpp
#include "services.h"
#include <fstream>
#include <iostream>
#include <chrono>

using namespace watcheD;
namespace services {

// This is exactly the same as in the cpu.lua file. This is here for memory usage difference only
class ServCpuCollector : public Collector {
public:
	ServCpuCollector(std::shared_ptr<HttpServer> p_srv, Json::Value* p_cfg, std::shared_ptr<service> p_serv): Collector("cpu", p_srv, p_cfg, 150, 10, p_serv) {
		addRessource("shed", "Service scheduler information", "service_cpu_shed");
		ressources["shed"]->addProperty("exec_runtime", "exec_runtime stat from the sheduler",	"number");
		ressources["shed"]->addProperty("process",	"Number of process",				"number");
		ressources["shed"]->addProperty("thread",	"total number of threads",			"number");
		addRessource("stat", "Service CPU usage", "service_cpu_stat");
		ressources["stat"]->addProperty("user",		"user %",		"number");
		ressources["stat"]->addProperty("system",	"system %",		"number");
		ressources["stat"]->addProperty("pct",		"total % used",		"number");
		ressources["stat"]->addProperty("cuser",	"user % cumulated",	"number");
		ressources["stat"]->addProperty("csystem",	"system % cumulated",	"number");
		addGetMetricRoute();
		prev_total_time=0;
	}

	void collect() {
		std::shared_ptr<service> serv_p = onService.lock();
		if (!serv_p) return;
		double sum		= 0;
		uint32_t process	= 0;
		uint32_t thread		= 0;
		uint32_t total_time	= 0;
		uint32_t cpu_count	= 0;
		uint32_t res_users	= 0;
		uint32_t res_sys	= 0;
		uint32_t res_cusers	= 0;
		uint32_t res_csys	= 0;
		std::string line;

		std::ifstream	f("/proc/stat");
		if (!f.good()) return;
		while(std::getline(f, line)) {
			std::istringstream iss(line);
			std::vector<std::string> tokens{std::istream_iterator<std::string>{iss}, std::istream_iterator<std::string>{}};
			if (tokens[0] == "cpu") {
				for(uint i=1;i<tokens.size();i++)
					total_time += atoi(tokens[i].c_str());
			} else if (tokens[0].substr(0,3) == "cpu")
				cpu_count++;
			
		}
		if (f.good()) f.close();
		if (cpu_count==0) cpu_count=1;

		std::shared_ptr< std::vector<uint32_t> > pids = serv_p->getPIDs();
		for (std::vector<uint32_t>::iterator i=pids->begin();i!=pids->end();i++) {
			// collect the shed information
			std::ifstream	infile("/proc/"+std::to_string(*i)+"/sched");
			if (!infile.good()) continue;
			while(std::getline(infile, line)) {
				std::istringstream iss(line);
				std::vector<std::string> tokens{std::istream_iterator<std::string>{iss}, std::istream_iterator<std::string>{}};
				if (tokens[0] == "se.sum_exec_runtime" ) {
					double v = stod(tokens[2]);
					if (values.find(*i)!=values.end() && values[*i]<v)
						sum += v - values[*i];
					values[*i] = v;
				} else if (tokens.size()>2 && tokens[2] == "#threads:")
					thread+=atoi(tokens[3].substr(0,tokens[3].length()-1).c_str());
			}
			infile.close();

			// collect cpu used data
			std::ifstream	statfile("/proc/"+std::to_string(*i)+"/stat");
			if (!statfile.good()) continue;
			if(std::getline(statfile, line)) { // there's should be only one line but...
				std::istringstream iss(line);
				std::vector<std::string> tokens{std::istream_iterator<std::string>{iss}, std::istream_iterator<std::string>{}};

				if (tokens.size() < 17) continue;
				uint32_t v = atoi(tokens[13].c_str());
				if (users.find(*i)!=users.end() && users[*i]<v)
					res_users += v - users[*i];
				users[*i] = v;
				v = atoi(tokens[14].c_str());
				if (sys.find(*i)!=sys.end() && sys[*i]<v)
					res_sys += v - sys[*i];
				sys[*i] = v;
				v = atoi(tokens[15].c_str());
				if (cusers.find(*i)!=cusers.end() && cusers[*i]<v)
					res_cusers += v - cusers[*i];
				cusers[*i] = v;
				v = atoi(tokens[16].c_str());
				if (csys.find(*i)!=csys.end() && csys[*i]<v)
					res_csys += v - csys[*i];
				csys[*i] = v;
			}
			statfile.close();
			process++;
		}
		if (prev_total_time!=0) {
			ressources["shed"]->nextValue();
			ressources["shed"]->setProperty("exec_runtime",	sum);
			ressources["shed"]->setProperty("process",	process);
			ressources["shed"]->setProperty("thread",	thread);
			double p = 100*cpu_count;p = p/(total_time-prev_total_time);
			ressources["stat"]->nextValue();
			ressources["stat"]->setProperty("user",		res_users*p);
			ressources["stat"]->setProperty("system",	res_sys*p);
			ressources["stat"]->setProperty("pct",		(res_sys+res_users)*p);
			ressources["stat"]->setProperty("cuser",	res_cusers*p);
			ressources["stat"]->setProperty("csystem",	res_csys*p);
		}
		prev_total_time = total_time;
	}
private:
	std::map<uint64_t, double> values;
	std::map<uint64_t, uint32_t> users;
	std::map<uint64_t, uint32_t> sys;
	std::map<uint64_t, uint32_t> cusers;
	std::map<uint64_t, uint32_t> csys;
	uint64_t prev_total_time;
};

MAKE_PLUGIN_SERVICE_COLLECTOR(ServCpuCollector, cpu)

}
